compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/DesktopTextInputService2.kt

```diff
@@ -44,19 +48,50 @@ internal class DesktopTextInputService2(
 
     private var composingText: String = ""
 
-    fun startInput(request: PlatformTextInputMethodRequest) {
+    private var ignoreNextInputMethodEvent = false
+
+    private var selectionListeningJob: Job? = null
+
+    fun startInput(request: PlatformTextInputMethodRequest, coroutineScope: CoroutineScope) {
         receivedInputMethodEventsSinceStartInput = false
+        ignoreNextInputMethodEvent = false
+
         component.enableInput(
             InputMethodSession(component, request).also {
                 inputMethodSession = it
             }
         )
+
+        // Listen to selection changes, and if a change doesn't match
+        // `inputMethodSession.expectedNextSelection`, end the composition
+        selectionListeningJob = coroutineScope.launch {
```

Comment:
```suggestion
        selectionListeningJob?.cancel()
        selectionListeningJob = coroutineScope.launch {
```
or `check(selectionListeningJob == null)`, but we should be sure that the `startInputMethod` coroutine can't be called in parallel by design.

Reply:
Sure.